topology2: Revisit period and buffer size settings

ujfalusi opened the pull request:
topology2: Revisit period and buffer size settings

To be pragmatic about the period/buffer size values:
period_size_min = 192 is for Stereo, S16_le, 48Khz, 1ms
period_size_max = 2097152 common value among other sound cards

The buffer size:
buffer_size_min	=	384 # period_size_min * periods_min
buffer_size_max	=	4194304 # period_size_max * 2

The deep buffer PCM only changes the buffer_size_min to (S16_LE, Stereo, 48KHz, DEEPBUFFER_FW_DMA_MS) * 2

plbossart commented on tools/topology/topology2/include/common/pcm_caps.conf:
```diff
@@ -88,8 +88,8 @@ Class.PCM."pcm_caps" {
-	buffer_size_min		65536
-	buffer_size_max		65536
+	buffer_size_min		384 # period_size_min * 2
+	buffer_size_max		-1
```
maybe use a more reasonable value? It's rather unlikely that the deep buffer will ever be more than 2s, that adds delays in user interaction to change tracks and apply volumes.

Infinity and beyond are not that useful for buffer sizes.

plbossart commented on tools/topology/topology2/include/common/pcm_caps.conf:
```diff
@@ -88,8 +88,8 @@ Class.PCM."pcm_caps" {
 	channels_min		2
 	channels_max		2
-	period_size_min		192
+	period_size_min		192 # Stereo, S16_LE, 48KHz, 1ms
```
why would we keep the period_size_max? Is there really any limitation?

ujfalusi commented:
Changes since v1:
- Be more pragmatic on the sizes and document them in commit message and in comments.

plbossart commented on tools/topology/topology2/include/common/pcm_caps.conf:
```diff
@@ -88,8 +88,8 @@ Class.PCM."pcm_caps" {
+	period_size_min		192 # Stereo, S16_LE, 48KHz, 1ms
+	period_size_max		19200 # Stereo, S16_LE, 48KHz, 100ms
+	buffer_size_min		384 # period_size_min * periods_min
+	buffer_size_max		307200 # period_size_max * periods_max
```
I would make this size up to 10s, and also account for higher resolutions/number of channels. If we need to support 8ch 192kHz the buffers are going to be quite small in the end,

ujfalusi replied:
For that we need to increase the periods_max.
These are the default parameters, if you override them then you need to update the sizes also.
In the definition we cannot use calculations, so it has to be done individually.

I can increase the periods_max to let's say 255?

ujfalusi replied:
10s for all PCMs? You still need to select either the number of periods or the period size max.
These are the default PCM parameters, not the DB ones.
We can have UINT_MAX for both buffer and period size max and call it a day. 1024 max periods and all config should work.

But we likely want to specify _some_ numbers, right? 1ms min for capture, 2ms for playback, 100ms as max period size as default. Number of periods? 255 should be enough.
The you re-calculate it for places where the default is changed (format, channels, rate)
DeeBuffer is a special one, we need 100ms as minimum (the kernel needs to help here), but what max period size we want? 1s? 500ms? 10s max buffer is fine, I guess but I don't want to over complicate things.

lgirdwood commented:
@ujfalusi any update here ? The pipeline base class should define teh default buffer sizes that can be overridden by derived classes. @ranj063 fyi.

ujfalusi commented:
Changes since v2:
- be more pragmatic and use a fixed period_size_max as 2097152.
- buffer_size_min = period_size_min * 2 (periods_min)
- buffer_size_max = period_size_max * 2 (twice as big)
- Deep Buffer only changes the buffer_size_min to avoid xrun on ALSA buffer (only with S16_LE, Stereo, 48K) as we have static value only

kv2019i reviewed (approved):
Good to merge @plbossart ?
